src/lib.rs

```diff
@@ -8647,6 +8654,68 @@ mod tests {
         assert_eq!(pipe.server.recovery.max_datagram_size(), 1200);
         assert_eq!(pipe.server.recovery.cwnd(), 12000);
     }
+
+    #[test]
+    /// Tests that connection-level send capacity decreases as more stream data
+    /// is buffered.
+    fn send_capacity() {
+        let mut buf = [0; 65535];
+
+        let mut config = Config::new(crate::PROTOCOL_VERSION).unwrap();
+        config
+            .load_cert_chain_from_pem_file("examples/cert.crt")
+            .unwrap();
+        config
+            .load_priv_key_from_pem_file("examples/cert.key")
+            .unwrap();
+        config
+            .set_application_protos(b"\x06proto1\x06proto2")
+            .unwrap();
+        config.set_initial_max_data(100000);
+        config.set_initial_max_stream_data_bidi_local(10000);
+        config.set_initial_max_stream_data_bidi_remote(10000);
+        config.set_initial_max_streams_bidi(10);
+        config.verify_peer(false);
+
+        let mut pipe = testing::Pipe::with_config(&mut config).unwrap();
+        assert_eq!(pipe.handshake(&mut buf), Ok(()));
+
+        assert_eq!(pipe.client.stream_send(0, b"hello!", true), Ok(6));
+        assert_eq!(pipe.advance(&mut buf), Ok(()));
+
+        assert_eq!(pipe.client.stream_send(4, b"hello!", true), Ok(6));
+        assert_eq!(pipe.advance(&mut buf), Ok(()));
+
+        assert_eq!(pipe.client.stream_send(8, b"hello!", true), Ok(6));
+        assert_eq!(pipe.advance(&mut buf), Ok(()));
+
+        assert_eq!(pipe.client.stream_send(12, b"hello!", true), Ok(6));
+        assert_eq!(pipe.advance(&mut buf), Ok(()));
+
+        let mut r = pipe.server.readable().collect::<Vec<u64>>();
+        assert_eq!(r.len(), 4);
+
+        r.sort();
+
+        assert_eq!(r, [0, 4, 8, 12]);
+
+        assert_eq!(pipe.server.stream_recv(0, &mut buf), Ok((6, true)));
+        assert_eq!(pipe.server.stream_recv(4, &mut buf), Ok((6, true)));
+        assert_eq!(pipe.server.stream_recv(8, &mut buf), Ok((6, true)));
+        assert_eq!(pipe.server.stream_recv(12, &mut buf), Ok((6, true)));
+
+        assert_eq!(pipe.server.tx_cap, 11565);
+
+        assert_eq!(pipe.server.stream_send(0, &buf[..5000], false), Ok(5000));
+        assert_eq!(pipe.server.stream_send(4, &buf[..5000], false), Ok(5000));
+        assert_eq!(pipe.server.stream_send(8, &buf[..5000], false), Ok(1565));
+
+        // No more connection send capacity.
+        assert_eq!(pipe.server.stream_send(12, &buf[..5000], false), Ok(0));
```

Comment:
`stream_send()` documentation says it will return `Done` if no data was written (https://github.com/cloudflare/quiche/blob/master/src/lib.rs#L2951). Is `Ok(0)` acceptable return value too?

Reply:
That's not been the case for a while AFAICT. In any case fix should go into a separate PR since it was the case before this change too.